components/support/nimbus-fml/src/intermediate_representation.rs

```diff
@@ -427,21 +433,36 @@ impl FeatureManifest {
     fn validate_feature_def(&self, feature_def: &FeatureDef) -> Result<()> {
         for prop in &feature_def.props {
             let path = format!("features/{}.{}", feature_def.name, prop.name);
-            self.validate_prop_defaults(&path, prop)?;
+            let literals = vec![prop.name.to_string()];
+            self.validate_default_by_typ(&path, &prop.typ, &prop.default, &literals)?;
         }
         Ok(())
     }
 
     fn validate_prop_defaults(&self, path: &str, prop: &PropDef) -> Result<()> {
-        self.validate_default_by_typ(path, &prop.typ, &prop.default)
+        let literals = Default::default();
+        self.validate_default_by_typ(path, &prop.typ, &prop.default, &literals)
     }
 
     pub fn validate_default_by_typ(
         &self,
         path: &str,
         type_ref: &TypeRef,
         default: &Value,
+        literals: &Vec<String>,
```

Comment:
This work in this method is to add the `literals` vec, a list of tokens leading from the root to where an error is detected.

The literals vec is then reported in errors, which are then unpacked in the `FmlClient`.

Several improvements suggest themselves:

- remove the `FMLError::ValidationError` from this method. Those problems are likely detected in another method.
- use the `preserve_order` feature of the `serde_json`, and add more to the `literals` vec. This will improve the accuracy of the error placement; however this should be done in a different issue— because it may need work in the `DefaultsMerger`.
- Process  the invalid property detection for features here, perhaps by translating the `FeatureDef` into `ObjectDef` (currently this is done in `DefaultsMerger`).
- Split this feature validation into a smaller more managable file/set of methods; this would make reviewing/editing the error messages easier, at the very least.

Reply:
Thank you so much for this explanation!! 🙏